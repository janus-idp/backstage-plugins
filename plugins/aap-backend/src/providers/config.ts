import { readSchedulerServiceTaskScheduleDefinitionFromConfig } from '@backstage/backend-plugin-api';
import type { Config } from '@backstage/config';

import type { AapConfig } from './types';

export function readAapApiEntityConfigs(config: Config): AapConfig[] {
  const providerConfigs = config.getOptionalConfig('catalog.providers.aap');
  if (!providerConfigs) {
    return [];
  }
  return providerConfigs
    .keys()
    .map(id => readAapApiEntityConfig(id, providerConfigs.getConfig(id)));
}

function readAapApiEntityConfig(id: string, config: Config): AapConfig {
  const baseUrl = config.getString('baseUrl');
  const authorization = config.getString('authorization');
  const system = config.getOptionalString('system');
  const owner = config.getOptionalString('owner') ?? 'unknown';

  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;

  return {
    id,
    baseUrl,
    authorization,
    system,
    owner,
    schedule,
  };
}
