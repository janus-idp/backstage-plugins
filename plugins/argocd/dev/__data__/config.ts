import { Config } from '@kubernetes/client-node';

export const mockArgocdConfig: Config = {
  argocd: {
    baseUrl: 'https://localhost:8080',
    appLocatorMethods: [
      {
        type: 'config',
        instances: [
          {
            name: 'argoInstance1',
            url: 'https://test-openshift-gitops.apps.test.devcluster.openshift.com/',
            token: 'fake-jwt-token1',
          },
          {
            name: 'argoInstance2',
            url: 'https://test-openshift-gitops.apps.test.devcluster.openshift.com/',
            token: 'fake-jwt-token2',
          },
        ],
      },
    ],
  },
};
