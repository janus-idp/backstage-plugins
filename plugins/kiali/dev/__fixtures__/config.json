{
  "server": {
    "accessibleNamespaces": ["**"],
    "authStrategy": "anonymous",
    "clusters": {
      "Kubernetes": {
        "apiEndpoint": "https://10.217.4.1:443",
        "isKialiHome": true,
        "kialiInstances": [
          {
            "namespace": "istio-system",
            "operatorResource": "kiali-operator/kiali",
            "serviceName": "kiali",
            "url": "https://kiali-istio-system.apps-crc.testing",
            "version": "dev"
          }
        ],
        "name": "Kubernetes",
        "network": "",
        "secretName": ""
      }
    },
    "deployment": {},
    "healthConfig": {
      "rate": [
        {
          "tolerance": [
            {
              "code": "5XX",
              "degraded": 0,
              "failure": 10,
              "protocol": "http",
              "direction": ".*"
            },
            {
              "code": "4XX",
              "degraded": 10,
              "failure": 20,
              "protocol": "http",
              "direction": ".*"
            },
            {
              "code": "^[1-9]$|^1[0-6]$",
              "degraded": 0,
              "failure": 10,
              "protocol": "grpc",
              "direction": ".*"
            },
            {
              "code": "^-$",
              "degraded": 0,
              "failure": 10,
              "protocol": "http|grpc",
              "direction": ".*"
            }
          ]
        },
        {
          "tolerance": [
            {
              "code": "5XX",
              "degraded": 0,
              "failure": 10,
              "protocol": "http",
              "direction": ".*"
            },
            {
              "code": "4XX",
              "degraded": 10,
              "failure": 20,
              "protocol": "http",
              "direction": ".*"
            },
            {
              "code": "^[1-9]$|^1[0-6]$",
              "degraded": 0,
              "failure": 10,
              "protocol": "grpc",
              "direction": ".*"
            },
            {
              "code": "^-$",
              "degraded": 0,
              "failure": 10,
              "protocol": "http|grpc",
              "direction": ".*"
            }
          ]
        }
      ]
    },
    "istioAnnotations": {
      "istioInjectionAnnotation": "sidecar.istio.io/inject"
    },
    "istioCanaryRevision": {},
    "istioConfigMap": "istio",
    "istioIdentityDomain": "svc.cluster.local",
    "istioLabels": {
      "appLabelName": "app",
      "injectionLabelName": "istio-injection",
      "injectionLabelRev": "istio.io/rev",
      "versionLabelName": "version"
    },
    "istioNamespace": "istio-system",
    "istioStatusEnabled": true,
    "kialiFeatureFlags": {
      "certificatesInformationIndicators": {
        "enabled": true,
        "secrets": ["cacerts", "istio-ca-secret"]
      },
      "clustering": { "enable_exec_provider": false },
      "istioAnnotationAction": true,
      "istioInjectionAction": true,
      "istioUpgradeAction": false,
      "uiDefaults": {
        "graph": {
          "findOptions": [
            {
              "description": "Find: slow edges (> 1s)",
              "expression": "rt > 1000"
            },
            {
              "description": "Find: unhealthy nodes",
              "expression": "! healthy"
            },
            {
              "description": "Find: unknown nodes",
              "expression": "name = unknown"
            },
            {
              "description": "Find: nodes with the 2 top rankings",
              "expression": "rank <= 2"
            }
          ],
          "hideOptions": [
            { "description": "Hide: healthy nodes", "expression": "healthy" },
            {
              "description": "Hide: unknown nodes",
              "expression": "name = unknown"
            },
            {
              "description": "Hide: nodes ranked lower than the 2 top rankings",
              "expression": "rank > 2"
            }
          ],
          "impl": "cy",
          "settings": {
            "fontLabel": 13,
            "minFontBadge": 7,
            "minFontLabel": 10
          },
          "traffic": { "grpc": "requests", "http": "requests", "tcp": "sent" }
        },
        "list": {
          "includeHealth": true,
          "includeIstioResources": true,
          "includeValidations": true,
          "showIncludeToggles": false
        },
        "metricsPerRefresh": "1m",
        "metricsInbound": {},
        "metricsOutbound": {},
        "refreshInterval": "60s"
      },
      "validations": {
        "ignore": ["KIA1301"],
        "SkipWildcardGatewayHosts": false
      }
    },
    "logLevel": "trace",
    "prometheus": {
      "globalScrapeInterval": 15,
      "storageTsdbRetention": 1296000
    }
  },
  "kialiConsole": "https://kiali-istio-system.apps-crc.testing/",
  "meshTLSStatus": {
    "status": "MTLS_NOT_ENABLED",
    "autoMTLSEnabled": true,
    "minTLS": "N/A"
  },
  "username": "anonymous",
  "istioCerts": [
    {
      "secretName": "istio-ca-secret",
      "secretNamespace": "istio-system",
      "dnsNames": null,
      "issuer": "O=cluster.local",
      "notBefore": "2023-08-28T06:39:18Z",
      "notAfter": "2033-08-25T06:39:18Z",
      "error": "",
      "accessible": true
    }
  ]
}
