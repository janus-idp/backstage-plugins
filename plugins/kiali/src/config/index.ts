export * from './Config';
export * from './ServerConfig';
export * from './Paths';
export * from './KialiLogo';
export * from './KialiIcon';
