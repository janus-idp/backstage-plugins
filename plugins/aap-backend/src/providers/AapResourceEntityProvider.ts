import { LoggerService } from '@backstage/backend-plugin-api';
import { PluginTaskScheduler, TaskRunner } from '@backstage/backend-tasks';
import {
  ANNOTATION_LOCATION,
  ANNOTATION_ORIGIN_LOCATION,
  Entity,
  ResourceEntity,
} from '@backstage/catalog-model';
import { Config } from '@backstage/config';
import {
  EntityProvider,
  EntityProviderConnection,
} from '@backstage/plugin-catalog-node';

import {
  listJobTemplates,
  listWorkflowJobTemplates,
} from '../clients/AapResourceConnector';
import { JobTemplate } from '../clients/types';
import { readAapApiEntityConfigs } from './config';
import { AapConfig } from './types';

export class AapResourceEntityProvider implements EntityProvider {
  private readonly env: string;
  private readonly baseUrl: string;
  private readonly authorization: string;
  private readonly owner: string;
  private readonly system: string;
  private readonly logger: LoggerService;
  private readonly scheduleFn: () => Promise<void>;
  private connection?: EntityProviderConnection;

  static fromConfig(
    configRoot: Config,
    options: {
      logger: LoggerService;
      schedule?: TaskRunner;
      scheduler?: PluginTaskScheduler;
    },
  ): AapResourceEntityProvider[] {
    const providerConfigs = readAapApiEntityConfigs(configRoot);

    return providerConfigs.map(providerConfig => {
      let taskRunner;
      if (options.scheduler && providerConfig.schedule) {
        taskRunner = options.scheduler.createScheduledTaskRunner(
          providerConfig.schedule,
        );
      } else if (options.schedule) {
        taskRunner = options.schedule;
      } else {
        throw new Error(
          `No schedule provided neither via code nor config for AapResourceEntityProvider:${providerConfig.id}.`,
        );
      }

      return new AapResourceEntityProvider(
        providerConfig,
        options.logger,
        taskRunner,
      );
    });
  }

  private constructor(
    config: AapConfig,
    logger: LoggerService,
    taskRunner: TaskRunner,
  ) {
    this.env = config.id;
    this.baseUrl = config.baseUrl;
    this.authorization = config.authorization;
    this.owner = config.owner;
    this.system = config.system || '';
    this.logger = logger.child({
      target: this.getProviderName(),
    });

    this.scheduleFn = this.createScheduleFn(taskRunner);
  }

  createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
    return async () => {
      const taskId = `${this.getProviderName()}:run`;
      return taskRunner.run({
        id: taskId,
        fn: async () => {
          try {
            await this.run();
          } catch (error: any) {
            // Ensure that we don't log any sensitive internal data:
            this.logger.error(
              `Error while syncing resources from AAP ${this.baseUrl}`,
              {
                // Default Error properties:
                name: error.name,
                message: error.message,
                stack: error.stack,
                // Additional status code if available:
                status: error.response?.status,
              },
            );
          }
        },
      });
    };
  }

  getProviderName(): string {
    return `AapResourceEntityProvider:${this.env}`;
  }

  async connect(connection: EntityProviderConnection): Promise<void> {
    this.connection = connection;
    await this.scheduleFn();
  }

  async run(): Promise<void> {
    if (!this.connection) {
      throw new Error('Not initialized');
    }

    this.logger.info(`Discovering ResourceEntities from AAP ${this.baseUrl}`);

    const entities: Entity[] = [];

    const templatesData = await Promise.allSettled([
      listJobTemplates(this.baseUrl, this.authorization),
      listWorkflowJobTemplates(this.baseUrl, this.authorization),
    ]);

    const templates: JobTemplate[] = [];

    templatesData.forEach(results => {
      if (results.status === 'fulfilled') {
        templates.push(...results.value);
      } else if (results.status === 'rejected') {
        // Ensure that we don't log any sensitive internal data:
        const error = results.reason || {};
        this.logger.error('Failed to fetch AAP job templates', {
          // Default Error properties:
          name: error.name,
          message: error.message,
          stack: error.stack,
          // Additional status code if available:
          status: error.response?.status,
        });
      }
    });

    for (const template of templates) {
      const resourceEntity: ResourceEntity =
        this.buildApiEntityFromJobTemplate(template);
      entities.push(resourceEntity);
      this.logger.debug(`Discovered ResourceEntity "${template.name}"`);
    }

    this.logger.info(`Applying the mutation with ${entities.length} entities`);

    await this.connection.applyMutation({
      type: 'full',
      entities: entities.map(entity => ({
        entity,
        locationKey: this.getProviderName(),
      })),
    });
  }

  private buildApiEntityFromJobTemplate(template: JobTemplate): ResourceEntity {
    const templateDetailsUrl = `${this.baseUrl}/#/templates/${template.type}/${template.id}/details`;
    const jobTemplateTransformedName = `${template.name.replace(/ /g, '_')}-${
      template.summary_fields?.organization?.name || template.id
    }-${this.env}`;

    return {
      kind: 'Resource',
      apiVersion: 'backstage.io/v1alpha1',
      metadata: {
        annotations: {
          [ANNOTATION_LOCATION]: this.getProviderName(),
          [ANNOTATION_ORIGIN_LOCATION]: this.getProviderName(),
        },
        name: `${jobTemplateTransformedName}`,
        title: `${template.name}`,
        description: `${template.description}`,
        links: [
          {
            url: `${this.baseUrl}`,
            title: 'AAP Dashboard',
          },
          {
            url: `${templateDetailsUrl}`,
            title: 'Template Details',
          },
        ],
      },
      spec: {
        type: `${template.type}`,
        owner: `${this.owner}`,
        ...(this.system && { system: `${this.system}` }),
      },
    };
  }
}
