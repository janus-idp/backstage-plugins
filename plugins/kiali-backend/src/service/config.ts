import { Config } from '@backstage/config';

const KIALI_PREFIX = 'catalog.providers.kiali';

export type KialiDetails = {
  url: string;
  skipTLSVerify?: boolean;
  sessionTime?: number;
  serviceAccountToken?: string;
  caData?: string;
  caFile?: string;
};

const isValidUrl = (url: string): boolean => {
  try {
    // eslint-disable-next-line no-new
    new URL(url);
  } catch (error) {
    return false;
  }
  return true;
};

export const getFromKialiConfig = (config: Config): Config => {
  // Check if required values are valid
  const requiredValues = ['url'];
  requiredValues.forEach(key => {
    if (!config.has(key)) {
      throw new Error(
        `Value must be specified in config at '${KIALI_PREFIX}.${key}'`,
      );
    }
  });
  return config;
};

export const getHubClusterFromConfig = (config: Config): KialiDetails => {
  const hub = getFromKialiConfig(config);

  const url = hub.getString('url');
  if (!isValidUrl(url)) {
    throw new Error(`"${url}" is not a valid url`);
  }

  return {
    url,
    serviceAccountToken: hub.getOptionalString('serviceAccountToken'),
    skipTLSVerify: hub.getOptionalBoolean('skipTLSVerify') || false,
    caData: hub.getOptionalString('caData'),
    caFile: hub.getOptionalString('caFile'),
    sessionTime: hub.getOptionalNumber('sessionTime'),
  };
};

export const readKialiConfigs = (config: Config): KialiDetails => {
  const kialiConfigs = config.getConfig(KIALI_PREFIX);
  return getHubClusterFromConfig(kialiConfigs);
};
