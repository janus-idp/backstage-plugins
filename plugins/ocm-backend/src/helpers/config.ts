import { Config } from '@backstage/config';
import { OcmConfig } from '../types';

const KUBERNETES_PLUGIN_CONFIG = 'kubernetes.clusterLocatorMethods';
const OCM_PREFIX = 'catalog.providers.ocm';
const KUBERNETES_PLUGIN_KEY = 'kubernetesPluginRef';

export const deferToKubernetesPlugin = (config: Config): boolean => {
  if (config.has(KUBERNETES_PLUGIN_KEY)) {
    return true;
  }
  return false;
};

export const getHubClusterFromKubernetesConfig = (
  id: string,
  config: Config,
  globalConfig: Config,
): Config => {
  const name = config.getOptionalString(KUBERNETES_PLUGIN_KEY);
  const _logTemplate = `Hub cluster ${OCM_PREFIX}.${id}.${KUBERNETES_PLUGIN_KEY}=${name}`;

  const hub = globalConfig
    .getConfigArray(KUBERNETES_PLUGIN_CONFIG)
    .flatMap(method => method.getOptionalConfigArray('clusters') || [])
    .find(cluster => cluster.getString('name') === name);
  if (!hub) {
    throw new Error(
      `${_logTemplate} not defined in kubernetes in ${KUBERNETES_PLUGIN_CONFIG}.clusters`,
    );
  }

  if (hub.getString('authProvider') !== 'serviceAccount') {
    throw new Error(`${_logTemplate} has to authenticate via 'serviceAccount'`);
  }
  return hub;
};

export const getHubClusterFromOcmConfig = (
  id: string,
  config: Config,
): Config => {
  // Check if required values are valid
  const requiredValues = ['name', 'url'];
  requiredValues.forEach(key => {
    if (!config.has(key)) {
      throw new Error(
        `Value must be specified in config at '${OCM_PREFIX}.${id}.${key}'`,
      );
    }
  });
  return config;
};

export const getHubClusterFromConfig = (
  id: string,
  config: Config,
  globalConfig: Config,
): OcmConfig => {
  const hub = deferToKubernetesPlugin(config)
    ? getHubClusterFromKubernetesConfig(id, config, globalConfig)
    : getHubClusterFromOcmConfig(id, config);

  return {
    id,
    url: hub.getString('url'),
    hubResourceName: hub.getString('name'),
    serviceAccountToken: hub.getOptionalString('serviceAccountToken'),
    skipTLSVerify: hub.getOptionalBoolean('skipTLSVerify') || false,
    caData: hub.getOptionalString('caData'),
  };
};

export const readOcmConfigs = (config: Config): OcmConfig[] => {
  const ocmConfigs = config.getOptionalConfig(OCM_PREFIX);

  if (!ocmConfigs) {
    return [];
  }

  return ocmConfigs
    .keys()
    .map(id => getHubClusterFromConfig(id, ocmConfigs.getConfig(id), config));
};
