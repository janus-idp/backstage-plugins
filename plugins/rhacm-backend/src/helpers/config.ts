import { Config } from '@backstage/config';

const CLUSTERS_PATH = 'kubernetes.clusterLocatorMethods';
const HUB_CLUSTER_CONFIG_PATH = 'rhacm.hub';

export const getHubClusterName = (config: Config) =>
  config.getString(HUB_CLUSTER_CONFIG_PATH);

export const getHubClusterFromConfig = (config: Config) => {
  try {
    getHubClusterName(config);
  } catch (err) {
    throw new Error(
      `Hub cluster must be specified in config at '${HUB_CLUSTER_CONFIG_PATH}'`,
    );
  }

  const hub = config
    .getConfigArray(CLUSTERS_PATH)
    .flatMap(method => method.getOptionalConfigArray('clusters') || [])
    .find(
      cluster =>
        cluster.getString('name') === config.getString(HUB_CLUSTER_CONFIG_PATH),
    );
  if (!hub) {
    throw new Error('Hub cluster not defined in kubernetes config');
  }
  return hub;
};
