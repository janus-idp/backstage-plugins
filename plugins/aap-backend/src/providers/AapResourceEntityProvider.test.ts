import { getVoidLogger } from '@backstage/backend-common';
import { TaskInvocationDefinition, TaskRunner } from '@backstage/backend-tasks';
import { ConfigReader } from '@backstage/config';
import { EntityProviderConnection } from '@backstage/plugin-catalog-node';

import {
  listJobTemplates,
  listWorkflowJobTemplates,
} from '../clients/AapResourceConnector';
import { AapResourceEntityProvider } from './AapResourceEntityProvider';

const AUTH_HEADER = 'Bearer xxxx'; // NOSONAR

const BASIC_VALID_CONFIG = {
  catalog: {
    providers: {
      aap: {
        dev: {
          baseUrl: 'http://localhost:8080',
        },
      },
    },
  },
} as const;

const BASIC_VALID_CONFIG_2 = {
  catalog: {
    providers: {
      aap: {
        dev: {
          baseUrl: 'http://localhost:8080',
          authorization: AUTH_HEADER,
        },
      },
    },
  },
} as const;

const connection = {
  applyMutation: jest.fn(),
  refresh: jest.fn(),
} as unknown as EntityProviderConnection;

jest.mock('../clients/AapResourceConnector', () => ({
  ...jest.requireActual('../clients/AapResourceConnector'),
  listJobTemplates: jest.fn().mockReturnValue({}),
  listWorkflowJobTemplates: jest.fn().mockReturnValue({}),
}));

class FakeAbortSignal implements AbortSignal {
  readonly aborted = false;
  readonly reason = undefined;
  onabort() {
    return null;
  }
  throwIfAborted() {
    return null;
  }
  addEventListener() {
    return null;
  }
  removeEventListener() {
    return null;
  }
  dispatchEvent() {
    return true;
  }
}

class ManualTaskRunner implements TaskRunner {
  private tasks: TaskInvocationDefinition[] = [];
  async run(task: TaskInvocationDefinition) {
    this.tasks.push(task);
  }
  async runAll() {
    const abortSignal = new FakeAbortSignal();
    for await (const task of this.tasks) {
      await task.fn(abortSignal);
    }
  }
  clear() {
    this.tasks = [];
  }
}

describe('AapResourceEntityProvider', () => {
  const logMock = jest.fn();

  const logger = getVoidLogger();
  logger.child = () => logger;
  ['log', ...Object.keys(logger.levels)].forEach(logFunctionName => {
    (logger as any)[logFunctionName] = function LogMock() {
      logMock(logFunctionName, ...arguments);
    };
  });

  const manualTaskRunner = new ManualTaskRunner();

  beforeEach(() => {
    jest.clearAllMocks();
    manualTaskRunner.clear();
  });

  afterEach(() => {
    const logs = JSON.stringify(logMock.mock.calls);
    // eslint-disable-next-line jest/no-standalone-expect
    expect(logs).not.toContain(AUTH_HEADER);
  });

  it('should return an empty array if no providers are configured', () => {
    const config = new ConfigReader({});

    const result = AapResourceEntityProvider.fromConfig(config, {
      logger,
    });

    expect(result).toEqual([]);
  });

  it('should not run without a authorization', () => {
    const config = new ConfigReader(BASIC_VALID_CONFIG);

    expect(() =>
      AapResourceEntityProvider.fromConfig(config, {
        logger,
      }),
    ).toThrow(
      "Missing required config value at 'catalog.providers.aap.dev.authorization",
    );
  });

  it('should not run without a valid schedule', () => {
    const config = new ConfigReader(BASIC_VALID_CONFIG_2);

    expect(() =>
      AapResourceEntityProvider.fromConfig(config, {
        logger,
      }),
    ).toThrow(
      'No schedule provided neither via code nor config for AapResourceEntityProvider:dev.',
    );
  });

  it('should return a single provider if one is configured', () => {
    const config = new ConfigReader(BASIC_VALID_CONFIG_2);
    const aap = AapResourceEntityProvider.fromConfig(config, {
      logger,
      schedule: manualTaskRunner,
    });

    expect(aap).toHaveLength(1);
  });

  it('should return provider name', () => {
    const config = new ConfigReader(BASIC_VALID_CONFIG_2);

    const aap = AapResourceEntityProvider.fromConfig(config, {
      logger,
      schedule: manualTaskRunner,
    });

    expect(aap.map(k => k.getProviderName())).toEqual([
      'AapResourceEntityProvider:dev',
    ]);
  });

  it('should connect', async () => {
    const config = new ConfigReader(BASIC_VALID_CONFIG_2);

    const aap = AapResourceEntityProvider.fromConfig(config, {
      logger,
      schedule: manualTaskRunner,
    });

    const result = await Promise.all(
      aap.map(async k => await k.connect(connection)),
    );
    expect(result).toEqual([undefined]);
  });

  it('should connect and run should resolves', async () => {
    (listJobTemplates as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          url: 'https://aap.com',
          name: 'demoJobTemplate',
          description: 'test description',
          type: 'job_template',
        },
      ]),
    );
    (listWorkflowJobTemplates as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          url: 'https://aap.worfkllow.com',
          name: 'demoWorkflowJobTemplate',
          description: 'test workflow description',
          type: 'workflow_job_template',
        },
      ]),
    );
    const config = new ConfigReader(BASIC_VALID_CONFIG_2);

    const aap = AapResourceEntityProvider.fromConfig(config, {
      logger,
      schedule: manualTaskRunner,
    });

    for await (const k of aap) {
      await k.connect(connection);
      await manualTaskRunner.runAll();
    }

    expect(connection.applyMutation).toHaveBeenCalledTimes(1);
    expect(
      (connection.applyMutation as jest.Mock).mock.calls,
    ).toMatchSnapshot();
  });

  it('should connect and run should resolves even if one api call fails', async () => {
    const error: Error & { config?: any; status?: number } = new Error(
      'Request failed with status code 401',
    );
    error.config = {
      header: {
        authorization: 'Bearer xxxx', // NOSONAR
      },
    };
    error.status = 401;
    (listJobTemplates as jest.Mock).mockRejectedValue(error);
    (listWorkflowJobTemplates as jest.Mock).mockReturnValue(
      Promise.resolve([
        {
          url: 'https://aap.worfkllow.com',
          name: 'demoWorkflowJobTemplate',
          description: 'test workflow description',
          type: 'workflow_job_template',
        },
      ]),
    );
    const config = new ConfigReader(BASIC_VALID_CONFIG_2);

    const aap = AapResourceEntityProvider.fromConfig(config, {
      logger,
      schedule: manualTaskRunner,
    });

    for await (const k of aap) {
      await k.connect(connection);
      await manualTaskRunner.runAll();
    }

    expect(connection.applyMutation).toHaveBeenCalledTimes(1);
    expect(
      (connection.applyMutation as jest.Mock).mock.calls,
    ).toMatchSnapshot();

    expect(logMock).toHaveBeenCalledWith(
      'info',
      'Discovering ResourceEntities from AAP http://localhost:8080',
    );
    expect(logMock).toHaveBeenCalledWith(
      'error',
      'Failed to fetch AAP job templates',
      {
        name: 'Error',
        message: 'Request failed with status code 401',
        stack: expect.any(String),
      },
    );
    expect(logMock).toHaveBeenCalledWith(
      'debug',
      'Discovered ResourceEntity "demoWorkflowJobTemplate"',
    );
  });
});
