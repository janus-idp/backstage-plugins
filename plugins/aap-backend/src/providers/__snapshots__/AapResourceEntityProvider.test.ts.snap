// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AapResourceEntityProvider should connect and run should resolves 1`] = `
[
  [
    {
      "entities": [
        {
          "entity": {
            "apiVersion": "backstage.io/v1alpha1",
            "kind": "Resource",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "AapResourceEntityProvider:dev",
                "backstage.io/managed-by-origin-location": "AapResourceEntityProvider:dev",
              },
              "description": "test description",
              "links": [
                {
                  "title": "AAP Dashboard",
                  "url": "http://localhost:8080",
                },
                {
                  "title": "Template Details",
                  "url": "http://localhost:8080/#/templates/job_template/undefined/details",
                },
              ],
              "name": "demoJobTemplate-undefined-dev",
              "title": "demoJobTemplate",
            },
            "spec": {
              "owner": "unknown",
              "type": "job_template",
            },
          },
          "locationKey": "AapResourceEntityProvider:dev",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1alpha1",
            "kind": "Resource",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "AapResourceEntityProvider:dev",
                "backstage.io/managed-by-origin-location": "AapResourceEntityProvider:dev",
              },
              "description": "test workflow description",
              "links": [
                {
                  "title": "AAP Dashboard",
                  "url": "http://localhost:8080",
                },
                {
                  "title": "Template Details",
                  "url": "http://localhost:8080/#/templates/workflow_job_template/undefined/details",
                },
              ],
              "name": "demoWorkflowJobTemplate-undefined-dev",
              "title": "demoWorkflowJobTemplate",
            },
            "spec": {
              "owner": "unknown",
              "type": "workflow_job_template",
            },
          },
          "locationKey": "AapResourceEntityProvider:dev",
        },
      ],
      "type": "full",
    },
  ],
]
`;

exports[`AapResourceEntityProvider should connect and run should resolves even if one api call fails 1`] = `
[
  [
    {
      "entities": [
        {
          "entity": {
            "apiVersion": "backstage.io/v1alpha1",
            "kind": "Resource",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "AapResourceEntityProvider:dev",
                "backstage.io/managed-by-origin-location": "AapResourceEntityProvider:dev",
              },
              "description": "test workflow description",
              "links": [
                {
                  "title": "AAP Dashboard",
                  "url": "http://localhost:8080",
                },
                {
                  "title": "Template Details",
                  "url": "http://localhost:8080/#/templates/workflow_job_template/undefined/details",
                },
              ],
              "name": "demoWorkflowJobTemplate-undefined-dev",
              "title": "demoWorkflowJobTemplate",
            },
            "spec": {
              "owner": "unknown",
              "type": "workflow_job_template",
            },
          },
          "locationKey": "AapResourceEntityProvider:dev",
        },
      ],
      "type": "full",
    },
  ],
]
`;
